include/zephyr/drivers/sensor_v2.h

```diff
@@ -0,0 +1,738 @@
+ *
+ * @see sensor_read_data() for argument description
+ */
+typedef int (*sensor_read_data_t)(const struct device *sensor, uint8_t sensor_type,
+				  struct sensor_data *data);
+
+/**
+ * @typedef sensor_get_range_t
+ * @brief Sensor API function for getting the current range
+ *
+ * @see sensor_get_range() for argument description
+ */
+typedef int (*sensor_get_range_t)(const struct device *sensor, uint8_t sensor_type, fp_t *range);
+
+/**
+ * @typedef sensor_set_range_t
+ * @brief Sensor API function for setting the current range
+ *
+ * @see sensor_set_range() for argument description
+ */
+typedef int (*sensor_set_range_t)(const struct device *sensor, uint8_t sensor_type, fp_t range,
+				  bool round_up);
+
+/**
+ * @typedef sensor_get_three_axis_bias_t
+ * @brief Sensor API function for getting the current bias and the temperature at which it was set
+ *
+ * @see sensor_get_three_axis_bias() for argument description
+ */
+typedef int (*sensor_get_three_axis_bias_t)(const struct device *sensor, uint8_t sensor_type,
+					    int8_t *temperature, fp_t *bias_x, fp_t *bias_y,
+					    fp_t *bias_z);
+
+/**
+ * @typedef sensor_set_three_axis_bias_t
+ * @brief Sensor API function for setting the bias and the temperature at which it was attained
+ *
+ * @see sensor_set_three_axis_bias() for argument description
+ */
+typedef int (*sensor_set_three_axis_bias_t)(const struct device *sensor, uint8_t sensor_type,
+					    int8_t temperature, fp_t bias_x, fp_t bias_y,
+					    fp_t bias_z);
+
+/**
+ * @typedef sensor_get_resolution_t
+ * @brief Sensor API function for getting the current resolution in bits/sample
+ *
+ * @see sensor_get_resolution() for argument description
+ */
+typedef int (*sensor_get_resolution_t)(const struct device *sensor, uint8_t sensor_type,
+				       uint8_t *resolution);
+
+/**
+ * @typedef sensor_set_resolution_t
+ * @brief Sensor API function for setting the current resolution in bits/sample
+ *
+ * @see sensor_set_resolution() for argument description
+ */
+typedef int (*sensor_set_resolution_t)(const struct device *sensor, uint8_t sensor_type,
+				       uint8_t resolution, bool round_up);
+
+/**
+ * @typedef sensor_process_data_callback_t
+ * @brief Callback type for handling asynchronous data
+ *
+ * This callback will be called once per data generated by the sensor in an asynchronous manner.
+ * Usually this means that the sensor is using a hardware FIFO and data was being read. There will
+ * be one call per data event.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param sensor_type The type of sensor that generated the data
+ * @param data The data generated, it should be cast to one of the sensor_*_data structs based on
+ *        the sensor_type.
+ * @return 0 on success
+ * @return < 0 on failure
+ *
+ * @see sensor_set_process_data_callback() for use.
+ */
+typedef int (*sensor_process_data_callback_t)(const struct device *sensor, uint8_t sensor_type,
```

Comment:
You're 100% right that this isn't obvious here and I'll work on setting up an example. I envisioned each sensor setting up their own buffer since FIFO size might vary. The idea would look something like:

```
/* Allocate a buffer that can hold 12 samples */
static uint8_t fifo_buffer[sizeof(struct sensor_data) + 11 * sizeof(struct sensor_sample_data)];

/* Cast the buffer so it's easier to access */
static struct sensor_data *fifo_sensor_data = (struct sensor_data*)fifo_buffer;

/* ISR handler */
{
  /* Get the timestamp and write it to fifo_sensor_data->header.base_timestamp */
}

/* Processing function */
{
  /* Reading N samples from the hardware fifo and write them to fifo_sensor_data->readings[i] */

  /* Set fifo_sensor_data->header.reading_count = N */

  /* Call the sensor_process_data_callback passing fifo_sensor_data which contains the full batch */
}
```

Now tying in your other comments. This does involve the copy from the sensor's hardware FIFO into the buffered sample, but the application doesn't need to do any copying. It can simply iterate over the `readings[]` of the data passed into the callback. Now this example is overly simple as it assumes only a single sensor type. Many sensors have for example both an accelerometer and gyroscope. In my thinking the driver would be responsible for splitting them up and calling the callback _N_ times depending on the presence of the sensor data of the FIFO.

Reply:
regarding the processing - reading from hardware and writing to readings will involve conversion to matching sensor_api type?